### DISCUSSION TITLE
Multi-select combobox losing selection state

### DISCUSSION THREAD

**Discussion by iamrakhmatov** [ariakit/ariakit#4104] [Questions] [answered]
We have a combobox component where user can search and we provide result based on the query, when user selects option and changes the search query and gets different results that doesn't include previously selected option, and when user keeps selecting another option from the result the previously selected option gets removed from the selection list on `setSelectedValue` callback.

**Comment by benrodrs** [comment]
> when user selects option and changes the search query and gets different results that doesn't include previously selected option

You have control over which values are rendered. Simply render `ComboboxItem` elements. If you're working on something similar to the [Multi-selectable Combobox](https://ariakit.org/examples/combobox-multiple), ensure `matches` includes `selectedValues`.

> when user keeps selecting another option from the result the previously selected option gets removed from the selection list on `setSelectedValue` callback.

I don't understand this part. If you think it's a bug, please [submit an issue with a reproduction](https://github.com/ariakit/ariakit/issues/new?assignees=&labels=potential+bug&projects=&template=bug.yml).

**Reply by iamrakhmatov** [reply]
yes, it's similar to Multi-select Combobox, but the thing is I can't ensure that all `selectedValues` are included in the `matches` because not all of the `selectedValues` can match for the search query user entered, some of them are selected when user searched for different results and selected from THAT `matches` so when query changes and correspondingly user gets new list of `matches` and selects from that, previous selections gets lost

**Reply by iamrakhmatov** [reply]
is there any way of I can store selections in the combobox state and even if I render different values and user selects from that I still can pass the selections from the combobox state alongside what was selected at the time?

**Reply by benrodrs** [reply]
> yes, it's similar to Multi-select Combobox, but the thing is I can't ensure that all `selectedValues` are included in the `matches` because not all of the `selectedValues` can match for the search query user entered, some of them are selected when user searched for different results and selected from THAT `matches` so when query changes and correspondingly user gets new list of `matches` and selects from that, previous selections gets lost

I don't get it. Can't you just change `matches` to include the missing `selectedValues`? Could you record a video or provide a reproduction of what you mean?

**Reply by iamrakhmatov** [reply]
https://github.com/user-attachments/assets/c449609e-0b9a-4724-8a51-48227597549c

**Reply by iamrakhmatov** [reply]
I can modify `matches` but it doesn't make sense because the previous selections may not match the query search user entered so we don't want to include them in the `matches` if they don't match the user search query but we want them to be preserved since they are user's selections from previous searches/matches

**Reply by benrodrs** [reply]
Thanks for the video. It seems you're doing something different from the [Multi-selectable Combobox](https://ariakit.org/examples/combobox-multiple). There, previously selected values are preserved after the list is filtered.

Without being able to reproduce it, I can't help much. However, I can take a look if you [submit an issue with a reproduction](https://github.com/ariakit/ariakit/issues/new?assignees=&labels=potential+bug&projects=&template=bug.yml).

**Reply by iamrakhmatov** [reply]
is there a way for multi-select combobox to preserve search query after selection? because right now in the example it's cleaning it from the input, I'm working on reproduction and I need this for it

**Reply by benrodrs** [reply]
Yes, set [`resetValueOnSelect`](https://ariakit.org/reference/combobox-item#resetvalueonselect) to `false` on `ComboboxItem`.

**Reply by iamrakhmatov** [reply, marked as answer]
@benrodrs I found the issue it was on our side, thank you for the help. closing the discuss
